### PR TITLE
Model Registry Settings: Add model registries table with db config viewer, delete action and stub for permission management modal

### PR DESCRIPTION
**Pull request by mturley**
Closes [RHOAIENG-6970](https://issues.redhat.com/browse/RHOAIENG-6970)
Closes [RHOAIENG-2230](https://issues.redhat.com/browse/RHOAIENG-2230) (followup to https://github.com/opendatahub-io/odh-dashboard/pull/2828)

## Description

* Adds the Model Registries table to the Model Registry Settings page
* Adds a stub modal for the "Manage Permissions" button, to be implemented in another story
* Adds a "View database configuration" kebab action which opens a modal. All details are displayed immediately except the password, which shows a loading skeleton while the associated secret is fetched.
  * For displaying the password in this modal, refactors the backend service so it handles secrets along with modelregistries
  * If there is an error fetching the associated secret or decoding the password from it, an inline alert is shown.
* Adds a "Delete model registry" kebab action with a modal requiring typing the name for confirmation.
* Using the updated backend service, adds `databasePassword` to the values passed when submitting the CreateModal

Note: Does not include the Database Status column from the mockups, that is blocked by backend work and has been moved into a followup story.

cc @yih-wang 

Table:
Screenshot 2024-05-20 at 5 18 13 PM

Kebab menu:
Screenshot 2024-05-21 at 11 51 03 AM

Password loading state:
Screenshot 2024-06-04 at 3 32 30 PM

Password loaded (clicking the eyeball shows it):
Screenshot 2024-05-20 at 5 20 34 PM

Password loading error state:
Screenshot 2024-05-21 at 11 46 20 AM

Delete modal:
Screenshot 2024-05-20 at 5 23 03 PM

Delete modal with name entered correctly:
Screenshot 2024-05-20 at 5 23 12 PM

## How Has This Been Tested?

Tested viewing existing model registries that were created on the shared cluster for testing, including viewing database details. Tested the password loading error by creating a model registry referencing a nonexistent secret and trying to view its details. Tested the delete operation.

## Test Impact

Cypress tests will be added in a separate PR, they'll conflict with tests @gitdallas has already started on

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by yih-wang** [comment]
@mturley Thanks for all the work! They look really great!

Can you use a plain alert for the case when the password cannot load? Additionally, is it possible for the user to successfully fetch the password with another attempt? If so, it might be beneficial to add a retry button here, like my second screenshot.
image
image

File: backend/src/routes/api/modelRegistrySettings/secrets/index.ts
Comment (andrewballantyne) [reply backend/src/routes/api/modelRegistrySettings/secrets/index.ts:7]: This should work like any other admin route -- not sure what I'm missing here, why the question?

**Comment by mturley** [comment]
Pushed the change to switch to inline alert and updated the screenshot above.

**Comment by mturley** [comment]
@lucferbux if you look at the Delete modal screenshots, am I correct in assuming that the default permission group shown in the message there (currently "TODO: group here") is something we can't fill in until we figure out [RHOAIENG-6636](https://issues.redhat.com/browse/RHOAIENG-6636) (RBAC story), or is that maybe known based on the registry name?

File: backend/src/routes/api/modelRegistrySettings/secrets/index.ts
Comment (mturley) [reply backend/src/routes/api/modelRegistrySettings/secrets/index.ts:7]: This is a route at `/api/modelRegistrySettings/secrets` which lets the user create, read, and update any secret name but hard-coded (via shared constant) to the odh-model-registries namespace. It was a compromise between having a wide open `/api/secrets` that would let a RHOAI admin mess with secrets in any namespace (security issue) and tightly coupling the secrets logic in the routes for reading/updating model registries themselves on the backend instead of doing it in the frontend (less flexible, harder to refactor later)

Comment (andrewballantyne) [reply backend/src/routes/api/modelRegistrySettings/secrets/index.ts:7]: What kind of flexibility do you need here? What kind of secrets are you modifying? @mturley

Comment (mturley) [reply backend/src/routes/api/modelRegistrySettings/secrets/index.ts:7]: The ModelRegistry spec has database credentials which include a `passwordSecret`, which is a reference to a Secret containing that db password. We have a create modal that needs to create a secret, then create a modelregistry referencing that secret, then take the uid of the new modelregistry and update the secret to add an ownerReference to it. Then we have a modal to view the database details including the password, so we need to look up the secret referenced by a modelregistry.

My initial proposal was that we modify our backend route for CRUD on modelregistries to take an object with `{ modelRegistry, password }` and handle the secret logic internally on the backend. But that felt a little restrictive since all the other backend services are pretty directly exposing API objects and the business logic is in the frontend, which I understand is good in case we need/want to pivot from having these backend routes and instead do all this through a passthrough API? @lucferbux mentioned something about it being more community-contribution-friendly this way as well, I can't recall the reasoning there.

**Comment by mturley** [comment]
/hold
let's rebase this after https://github.com/opendatahub-io/odh-dashboard/pull/2828 and implement the Secret logic part of submitting that create modal as part of this PR. That PR also establishes cypress tests that this PR can add to.

File: backend/src/routes/api/modelRegistries/index.ts
Comment (gitdallas) [inline backend/src/routes/api/modelRegistries/index.ts:55]: if `dryRun` has a "string" type, could it be "false"? If so, should this be changed to account for that?

**Comment by mturley** [comment]
/unhold

@lucferbux this is ready for review now, sorry for the delay on the cypress tests!

File: backend/src/routes/api/modelRegistries/index.ts
Comment (mturley) [reply backend/src/routes/api/modelRegistries/index.ts:55]: In the k8s API, `dryRun` can only either be undefined or `"All"`. It's a string in case future API versions add more types of dry-runs I guess.

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/
> When present, indicates that modifications should not be persisted. An invalid or unrecognized dryRun directive will result in an error response and no further processing of the request. Valid values are: - All: all dry run stages will be processed

We could maybe use a more specific type `"All" | undefined` on our end, but the fastify methods we're calling expect a plain `string`.

**Comment by mturley** [comment]
/hold
Dallas found an issue I'm looking into

**Comment by mturley** [comment]
/unhold
False alarm. Need to test with local backend instead of `start:dev:ext` since the changes aren't on the cluster

**Review by gitdallas — APPROVED** [review]

**Comment by mturley** [comment]
Pushed changes to address some feedback after a Slack conversation with @gitdallas:
* When creating or deleting a registry, the modal does not close until the list of registries has been fully refreshed (so we no longer get a lag between the submit completing and the list of registries changing).
* When viewing database details, instead of showing a one-line loading skeleton while the password loads, the whole modal shows a spinner. The skeleton component seemed potentially confusing (if it's taking a while, it could just look like an intentionally obscured password). If there is an error loading the password, the rest of the database details will still be shown. Updated the loading state screenshot in the description above.

**Review by gitdallas — APPROVED** [review]

File: backend/src/routes/api/modelRegistries/modelRegistryUtils.ts
Comment (lucferbux) [inline backend/src/routes/api/modelRegistries/modelRegistryUtils.ts:14]: Mmmm yeah, great idea having the utility function, we've always encoded/decoded without having a declared function in the backend

File: frontend/src/pages/modelRegistrySettings/CreateModal.tsx
Comment (lucferbux) [inline frontend/src/pages/modelRegistrySettings/CreateModal.tsx:29]: Mmmmm didn't comment this before, and it's not a bad addition cause in the admin panel we were using old not refactored forms, but we got `useGenericObjectState` for creating objects and using them for forms.

We should refactor this modal in the near future with a custom hook that feeds from `useGenericObjectState`

File: frontend/src/pages/modelRegistrySettings/CreateModal.tsx
Comment (lucferbux) [inline frontend/src/pages/modelRegistrySettings/CreateModal.tsx:92]: Remind me of talking about the posibility of using descriptive names for model registries, we sholud have a display name option, but we should talk with ux first.

**Review by lucferbux — APPROVED** [review]
/lgtm

**Comment by lucferbux** [comment]
Code looks great, there are a few open questions that i wanna discuss with you @mturley so we can do follow ups, but so far, it can be merged!

**Comment by lucferbux** [comment]
(feel free to resolve all comments once you read them so we can proceed merging the pr)

**Comment by mturley** [comment]
Looks like the bot is happy to merge with unresolved conversations now 🤷 

@lucferbux should we open followup issues for the 2 points you raised above (refactoring to use `useGenericObjectState` and supporting display names for registries)?

**Pull request merged**
